Add mqtt-zenoh reference implementation and mqtt example entities (#58)
eclipse-uprotocol/up-streamer-rust: example-streamer-implementations/README.md

# up-linux-streamer

These are reference implementations of a uStreamer.
We implement the Streamers so that they forward between two components, either between Zenoh and MQTT or between Zenoh and SomeIP.
A component running with Zenoh could be a component running in an ECU, one running with MQTT could be a component running in some sort of cloud and one running SomeIP could be a mechatronics component.
The Streamer by itself does not do much, but its designed to be run together with two of the transport examples in the example-streamer-uses folder.

## Supported Setups

Here are the setups that can be built with these streamers and different entities with the Zenoh, SomeIP and MQTT transports.
Depending on which setup you want to try, run the streamer with:

```bash

> cargo run --bin <zenoh_mqtt/zenoh_someip> --features=<check cargo.toml or logs to see which features you need> -- --config="./example-streamer-implementations/README.md"

Author: fixed

The subscriber listens to one or multiple topics via a filter.

For this setup you can also choose:
- a Zenoh Publisher and an MQTT Subscriber (A backend service getting live data from a car)
- an MQTT Publisher and a Zenoh Subscriber (A car getting over the air traffic information)
- a Zenoh Publisher and a SomeIP Subscriber (An autoedge app pushing some configurations to a mechatronics component)
- or a SomeIP Publisher and a Zenoh Subscriber (An infotainment app getting live data from a sensor)

## Understanding the Configuration Files

Reference the `DEFAULT_CONFIG.json5` configuration file to understand the basic configuration options for the Streamer.

The `ZENOH_CONFIG.json5` file is used to set Zenoh configurations. By default, it is only used to set listening endpoints, but can be used with more configurations according to [Zenoh's page on it](https://zenoh.io/docs/manual/configuration/#configuration-files).

The 'static_subscriptions.json' is only needed when you set up a publish-subscribe system and can be ignored for a client-service system.

> The 'vsomeip-config/point_to_point.json'

Author: done

> The MQTT Transport has its configuration hardcoded for now. For the settings check main.rs for "mqtt_config".

Reviewer: Are there pieces that make sense to expose here now? I'm imagining if/when folks try to use this reference streamer for the Eclipse SDV Hackathon it could make it easier.

> ## Running the Streamer in system with an ECU that supports Zenoh and a cloud component that runs MQTT

Reviewer: Could you add a section like this, but for the SOME/IP & Zenoh setup?

Author: done!

### Running the `up-linux-streamer-mqtt-zenoh` as the Streamer instance

To run one of the examples below and see client and service communicate, you'll need to run the `up-linux-streamer-mqtt-zenoh` to bridge between the transports in a terminal. This implementation works out of the box with the given examples.

First run an MQTT broker. You can use your own or use the Mosquitto instance provided in this repo. For that execute the following:

```bash
cd mosquitto
docker compose up
```

Run the linux streamer with the default configuration file from the project root directory (or fix the path and run from here):

```bash

> cargo run -p up-linux-streamer-mqtt-zenoh -- --config up-linux-streamer-mqtt-zenoh/DEFAULT_CONFIG.json5

Reviewer: ```suggestion
cargo run -p zenoh_mqtt -- --config up-linux-streamer-mqtt-zenoh/DEFAULT_CONFIG.json5
```
?

```

This starts the streamer which should now be idle. As soon as a client tries to connect with the streamer, the connection will be logged.

### Running the Entities

Execute the following command from the project root directory to start one of the Entities:

```bash
cargo run -p example-streamer-uses --bin <transport_entity>
```

> Depending on the setup you want to test, chose any of these examples:

Reviewer: ```suggestion
Depending on the setup you want to test, chose any of these <transport_entity>:
```

Author: fixed I think

| Entity 1        | Entity 2         |
| --------------- | -------------    |
| mqtt_client     | zenoh_service    |
| mqtt_service    | zenoh_client     |
| mqtt_publisher  | zenoh_subscriber |
| mqtt_subscriber | zenoh_publisher  |

> The service and client will run forever. Every second a new request message is sent from the client via zenoh. That zenoh message is caught and routed over MQTT to the service. The response to the request makes the same journey in reverse.

Reviewer: ```suggestion
The service and client will run forever. Every second a new request message is sent from the client via Zenoh. That zenoh message is caught and routed over MQTT to the service. The response to the request makes the same journey in reverse.
```

Capitalize Zenoh

Author: done

| mqtt_publisher  | zenoh_subscriber |
| mqtt_subscriber | zenoh_publisher  |

The service and client will run forever. Every second a new request message is sent from the client via zenoh. That Zenoh message is caught and routed over MQTT to the service. The response to the request makes the same journey in reverse.

## Running the Streamer in a system consisting of an ECU that supports Zenoh and a mechatronics component using SOME/IP

### Running the `zenoh_someip` binary as the Streamer instance

The other available reference implementation streams between the Zenoh and SOME/IP transports.

For this implementation you need to add the path to your vsomeip library to the environment, for example by setting it in your .cargo/config.toml:

```toml
[env]
VSOMEIP_INSTALL_PATH=<path to your vsomeip lib (for example /usr/local/lib)>
```

Run the "zenoh_someip" binary of the streamer with the default configuration file:

```bash

> cargo run -bin zenoh_someip --features="zenoh-transport vsomeip-transport" -- --config="DEFAULT_CONFIG.json5"

Reviewer: ```suggestion
cargo run -bin zenoh_someip --features="zenoh-transport vsomeip-transport bundled-someip" -- --config="DEFAULT_CONFIG.json5"
```

Could also note the ability to include the `bundled-vsomeip` feature flag to avoid needing to compile your own vsomeip.

Let's do this in a follow-up PR so we can merge what you've done so far! ❤️
